ocaml/runtime/signals_nat.c

```diff
@@ -85,3 +85,39 @@ void caml_garbage_collection(void)
                               nallocs, alloc_len);
   }
 }
+
+#define DECLARE_SIGNAL_HANDLER(name) \
+  static void name(int sig, siginfo_t * info, ucontext_t * context)
+
+#define SET_SIGACT(sigact,name)                                       \
+  sigact.sa_sigaction = (void (*)(int,siginfo_t *,void *)) (name);    \
+  sigact.sa_flags = SA_SIGINFO
+
+#include <unistd.h>
+
+DECLARE_SIGNAL_HANDLER(segv_handler)
+{
+  struct sigaction act;
+  struct stack_info *block = Caml_state->current_stack;
+  char* fault_addr = info->si_addr;
+  int page_size = getpagesize();
+  char* protected_low = (char *) block + page_size;
+  char* protected_high = protected_low + page_size;
+  if ((fault_addr >= protected_low) && (fault_addr < protected_high)) {
+    caml_raise_stack_overflow();
```

Comment:
This looks suspicious to me - you're calling the C code to raise an exception, but you're probably on the OCaml stack. Have you tried catching this exception with Sys.with_async_exns? Does it work?

Also, I don't think you're guaranteed to hit the fault at function entry, since it doesn't look like functions touch their whole stackframe early. That may mean that something goes wrong when trying to compute a backtrace here.

Returning from a signal handler like this is dicey and platform-dependent. See the ifdef'd code in OCaml 4 for details.

Reply:
Changed to the OCaml 4 code.